src/models/collection.ts

```diff
@@ -1,14 +1,28 @@
 import type { BaseModel } from "./base";
+import type { DetailedAsyncAPI } from "../types";
 
-export abstract class Collection<T extends BaseModel | Collection<any>> extends Array<T> {
+export interface CollectionMetadata<T = any> {
+  originalData?: Record<string, T>;
+  asyncapi?: DetailedAsyncAPI;
+  pointer?: string;
+}
+
+export abstract class Collection<T extends BaseModel = BaseModel, M extends Record<string, any> = {}> extends Array<T> {
   constructor(
-    protected readonly collections: T[]
+    protected readonly collections: T[],
+    protected readonly _meta: CollectionMetadata<T> & M = {} as CollectionMetadata<T> & M,
   ) {
     super(...collections);
   }
 
-  abstract get(id: string): T | undefined;
-  abstract has(id: string): boolean;
+  get(id: string): T | undefined {
+    const possibleItem = this._meta.originalData?.[id];
+    return typeof possibleItem === 'undefined' ? this.__get(id) : possibleItem;
```

Comment:
Shouldn't this be just the opposite? First we see if `__get` is declared, if so, means there is a desire from the subclass to override the whole `get`?

Reply:
Good point, but have in mind that `__get` here is treated as fallback to retrieve data when `originalData` (which can be omitted) doesn't exist or given key also doesn't exist. With your solution we need to duplicate `const possibleItem = this._meta.originalData?.[id];` logic in every `__get` function.

Reply:
What I suggest is that we do the opposite; something like:

```js
const possibleItem = this.__get(id);
return typeof possibleItem === 'undefined' ? this._meta.originalData?.[id] : possibleItem;
```

Reply:
But by this you will decrease optimization, I made looking for object first to improve read speed. Also `originalData` is optional so we can pass objects to the collections constructor only when we will have 100% sure that keys in objects === ids in collections.

Reply:
Correct me if I'm wrong, please. Imagine having the following document:

```yaml
components:
  messages:
    test1:
      messageId: 'test'
    test2:
      messageId: 'test1'
```

What '.components().messages().get('test1')` should return? 
I expect `messageId` should have preference over the key, so I would say the returned message would be:

```yaml
test2:
      messageId: 'test1'
```

However, with your change, that won't happen since `this._meta.originalData?.[id]` will match with the other message.

Reply:
Nope, as I wrote:

```
Also `originalData` is optional so we can pass objects to the collections constructor only when we will have 100% sure that keys in objects === ids in collections.
```

so in messages/operations cases where we use `operationId` and `messageId`, we will always run `__get`.

Reply:
Being hones, I'm not very convinced about the complexity of that code. Regarding the fact `originalData` is optional, it is like a hidden behavior that it is not very clear to the developer what happens when it is provided. Not even convinced it would be fixed by documenting.
Is it worth the "optimization"?

Reply:
We wrote that there is a problem with optimization the reading of data as an array, and it would be nice to have a way to speed this up using an object and keys. As for the complexity of this, it is only for internal use. The end user should not create standalone models at all that are not AsyncAPIDocument model.

Reply:
I can remove that if you want, because atm we only use it in the `extensions` case for bindings. WDYT?

Reply:
I'm concerned about the Developer (of the parser) Experience, not the final user. 
I understand the benefit; kind of o(1) VS o(n) lookups, However I'm always in favor of not introducing early optimizations in code, and this one might be one, especially if not executed thinking on DX.

As a last effort to keep it, maybe at least renaming the `originalData` var to `indexedCollection` or similar and adding a comment on what this is about could help. 

Just to clarify, this could also be used for example for `servers` and `channels` (v2) right?

Reply:
I opt for `originalData` because in bindings model we need to pass original data (not changed), because then we need to read from original `bindings` field all extensions for `extensions()` method.  `indexedCollection` indicates that this data is transformed. 

> Just to clarify, this could also be used for example for servers and channels (v2) right?

Yeah, also for bindings and extensions models - in every collections where we change object to array 1:1 - we change key to the id of element.

I will remove __get, but we need `originalData` meta field.

Reply:
`indexedCollection` doesn't indicate that data is transformed but just indexed by something (we can assume by id).

Reply:
Yeah, but we can even transform to the object original data - for example messages we can transform to the object, where keys will be `messageId`s.

Ok, I removed the `__get` function. We will think about optimization after release.